lib/pg_ha_migrations/safe_statements.rb

```diff
@@ -170,6 +170,65 @@ def safe_remove_concurrent_index(table, options={})
     unsafe_remove_index(table, **options.merge(:algorithm => :concurrently))
   end
 
+  def safe_add_concurrent_partitioned_index(table, columns, name_suffix: nil, if_not_exists: false)
+    if ActiveRecord::Base.connection.postgresql_version < 11_00_00
+      raise PgHaMigrations::InvalidMigrationError, "Concurrent partitioned index creation not supported on Postgres databases before version 11"
+    end
+
+    schema, table = _schema_and_table_for(table)
+
+    is_partitioned = select_value(<<~SQL)
+      SELECT EXISTS (
+        SELECT 1
+        FROM pg_partitioned_table, pg_class, pg_namespace
+        WHERE pg_class.oid = pg_partitioned_table.partrelid
+          AND pg_class.relnamespace = pg_namespace.oid
+          AND pg_class.relname = #{quote(table)}
+          AND pg_namespace.nspname = #{quote(schema)}
+      )
+    SQL
+
+    raise PgHaMigrations::InvalidMigrationError, "Table #{table} is not a partitioned table" unless is_partitioned
+
+    name_suffix = "#{Array.wrap(columns).join("_")}_idx" unless name_suffix.present?
+
+    child_tables = select_values(<<~SQL)
+      SELECT child.relname
+      FROM pg_inherits
+        JOIN pg_class parent ON pg_inherits.inhparent = parent.oid
+        JOIN pg_class child  ON pg_inherits.inhrelid  = child.oid
+        JOIN pg_namespace    ON parent.relnamespace = pg_namespace.oid
+      WHERE parent.relname = #{quote(table)}
+        AND pg_namespace.nspname = #{quote(schema)}
+    SQL
+
+    parent_index = "#{table}_#{name_suffix}"
+
+    quoted_columns = Array.wrap(columns).map { |c| quote_column_name(c) }.join(", ")
+
+    unsafe_execute(<<~SQL)
+      CREATE INDEX#{if_not_exists ? " IF NOT EXISTS" : ""} #{quote_column_name(parent_index)}
+      ON ONLY #{quote_schema_name(schema)}.#{quote_table_name(table)} (#{quoted_columns})
+    SQL
+
+    child_tables.each do |child_table|
+      child_index = "#{child_table}_#{name_suffix}"
+
+      next if if_not_exists && _index_exists?(schema, child_index)
+
+      safe_add_concurrent_index("#{quote_schema_name(schema)}.#{child_table}", columns, name: child_index)
+
+      unsafe_execute(<<~SQL)
+        ALTER INDEX #{quote_schema_name(schema)}.#{quote_column_name(parent_index)}
+        ATTACH PARTITION #{quote_schema_name(schema)}.#{quote_column_name(child_index)}
+      SQL
```

Comment:
This should be relatively quick but it takes out an access exclusive lock. Should we wrap this in a call to `adjust_lock_timeout` or `adjust_statement_timeout`?

Reply:
We need to wrap it in `safe_acquire_lock_for_table` (and a `say_with_time` I think).

Reply:
So apparently whatever website I was looking at was wrong (I can't even find it at this point). After further investigation, on PG 11-13 `ALTER INDEX ... ATTACH PARTITION` takes out an `AccessShareLock` on the child table and on PG 14-15 it takes out `AccessShareLock` on the child _and_ parent table (or at least that's what is being displayed in the `pg_locks` view).

Considering issue https://github.com/braintree/pg_ha_migrations/issues/39 hasn't been implemented yet, I don't think we want `safe_acquire_lock_for_table` here

Reply:
> What should the message be for say_with_time though?

I realize we don't have a method name it can output, but it seems like it's still an operation we should expose (it is DDL after all); maybe something like `Attaching per-partition index #{index name} to parent #{index name}`?

Reply:
> After further investigation, on PG 11-13 ALTER INDEX ... ATTACH PARTITION takes out an AccessShareLock on the child table and on PG 14-15 it takes out AccessShareLock on the child and parent table (or at least that's what is being displayed in the pg_locks view).

We need to do more verification (and perhaps code spelunking). According to the [Postgres docs](https://www.postgresql.org/docs/current/sql-alterindex.html):

> the lock level required may differ for each subform. An ACCESS EXCLUSIVE lock is held unless explicitly noted.

and the `ATTACH PARTITION` subform has no such note. That being said it would very much surprise me to find that it's doing a lock escalation, since that usually is a vector for deadlocks.

Reply:
We talked about this one offline as well:

Not sure why I wasn't able to observe the `AccessExclusiveLock` in testing, but the source code makes it pretty clear what's happening: https://github.com/postgres/postgres/blob/c0dc97c7bfd14b460536c40d14787f82ab61d0d9/src/backend/commands/tablecmds.c#L18981

The latest commit wraps all of the `ATTACH PARTITION` calls in a `safely_acquire_lock_for_table` block